More comprehensive documentation for CSRF protection (#906)
vitalik/django-ninja: docs/docs/reference/csrf.md

# CSRF

## What is the CSRF protection and why is it needed?

> Cross-site request forgery is a type of vulnerabilities where a user's web browser connects to a malicious page, which triggers request to your backend (using hidden forms, JavaScript `fetch` or `XMLHttpRequest`, or other methods), that could interpret those as legitimate as some credentials saved on the user's web browser (like cookies) may be automatically embedded.

Reviewer: ```suggestion
Cross-site request forgery is a type of vulnerability where a user's web browser connects to a malicious page, which triggers request to your backend (using hidden forms, JavaScript `fetch` or `XMLHttpRequest`, or other methods), that could interpret those as legitimate as some credentials saved on the user's web browser (like cookies) may be automatically embedded.
```

Your backend must be able to differentiate between legitimate and malicious requests. This can be done in several ways.

> If you are not familiar with CSRF attacks it is recommended to at leat read:

Reviewer: ```suggestion
If you are not familiar with CSRF attacks it is recommended to at least read:
```

> ## What is the CSRF protection and why is it needed?
> Cross-site request forgery is a type of vulnerabilities where a user's web browser connects to a malicious page, which triggers request to your backend (using hidden forms, JavaScript `fetch` or `XMLHttpRequest`, or other methods), that could interpret those as legitimate as some credentials saved on the user's web browser (like cookies) may be automatically embedded.
> 
> Your backend must be able to differentiate between legitimate and malicious requests. This can be done in several ways.
> 
> If you are not familiar with CSRF attacks it is recommended to at leat read:
> - the [Wikipedia article about CSRF](https://en.wikipedia.org/wiki/Cross-site_request_forgery)
> For a more complete understanding of different mitigations and their implications, you may also read:
> - the [Owasp Community Page about CSRF](https://owasp.org/www-community/attacks/csrf)
> - the [OWASP Cheat Sheet about Cross-Site Request Forgery Prevention](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html)
> 
> 
> ## How to protect with Django Ninja
> ### Use an authentication method that doesn't get automatically transmitted by request from another site
> CSRF can happen because the cookies are automatically included in requests started from another site.
> Not using cookies but another method to transfer and consume the token, like the `Authorization: Bearer` header for exemple, mitigates this attack.
> 
> ### Django CSRF protection
> In case you are using the default Django authentication, which uses cookies, you must also use the default [Django CSRF protection](https://docs.djangoproject.com/en/4.2/ref/csrf/).

Reviewer: It might be better to rely on more heavily on the django documentation, rather than providing our own security advice.

```suggestion
## What is the CSRF?
> [Cross Site Request Forgery](https://www.squarefree.com/securitytips/web-developers.html#CSRF) occurs when a malicious website contains a link, a form button or some JavaScript that is intended to perform some action on your website, using the credentials of a logged-in user who visits the malicious site in their browser. A related type of attack, ‘login CSRF’, where an attacking site tricks a user’s browser into logging into a site with someone else’s credentials, is also covered.

## How to protect against CSRF with Django Ninja
### Use a non-cookie-based authentication method
CSRF attacks rely on cookies that are automatically included in requests started from another site.
Using an authentication method that does not rely on cookies, such as the `Authorization: Bearer` header for exemple, mitigates this attack.

### Use Django's built-in CSRF protection
In case you are using the default Django authentication, which uses cookies, you must also use the default [Django CSRF protection](https://docs.djangoproject.com/en/4.2/ref/csrf/).
```

Author: I can't commit your suggestion anymore but I did integrate most of it, don't hesitate to unresolve if it is unclear @OtherBarry

from ninja.security import django_auth

api = NinjaAPI(auth=django_auth)
```

> You can use the Django [ensure_csrf_cookie](https://docs.djangoproject.com/en/4.2/ref/csrf/#django.views.decorators.csrf.ensure_csrf_cookie) decorator on an unprotected route to make it include a `Set-Cookie` header for the CSRF token (note that the route decorator must come upper than the [ensure_csrf_cookie](https://docs.djangoproject.com/en/4.2/ref/csrf/#django.views.decorators.csrf.ensure_csrf_cookie) decorator).

Reviewer: ```suggestion
You can use the Django [ensure_csrf_cookie](https://docs.djangoproject.com/en/4.2/ref/csrf/#django.views.decorators.csrf.ensure_csrf_cookie) decorator on an unprotected route to make it include a `Set-Cookie` header for the CSRF token (note that the route decorator must be executed before (i.e. above) the [ensure_csrf_cookie](https://docs.djangoproject.com/en/4.2/ref/csrf/#django.views.decorators.csrf.ensure_csrf_cookie) decorator).
```

Reviewer 2: Are you sure what is described here works? In my experience, no decorator doing work in `process_response` works(ref #544).

This might very well have changed. But is there perhaps e.g. a test that could be pointed to?

Author: Yes, it does work, but only with `HttpResponse`, not a dict. It is true it is unclear, I will update my MR later.
